src/main/java/org/aksw/gerbil/dataset/impl/conll/GenericCoNLLDataset.java

```diff
@@ -193,71 +205,291 @@ protected List<Document> loadDocuments(File file) throws GerbilException {
+                    break;
+                }
+                case '፡': // ፡ word separator
+                default: {
+                    whiteSpaceInFront = true;
+                    whiteSpaceBehindCurrentToken = true;
+                    break;
+                }
+                }
+//                } else if (!Character.isLetterOrDigit(token[0].charAt(0))) {
+//                    // Check if the first character of the line is not a letter or digit
+//                    whiteSpaceInFront = false;
+//                    // Set whiteSpaceInFront to false if the line starts with a non-letter or
+//                    // non-digit character
+            } else {
+                whiteSpaceInFront = true;
+                whiteSpaceBehindCurrentToken = true;
+            }
+            // Remove leading/trailing whitespaces and normalize spaces within the token
+            String normalizedToken = token[0].trim().replaceAll("\\s+", " ");
+
+            // If the current marking is not null AND there is no annotation column or there
+            // is no MARKING_INSIDE annotation --> The previous marking ended
+            if (currentMarking != null
+                    && (token.length <= annotationColumn || !token[annotationColumn].startsWith(MARKING_INSIDE))) {
+                currentMarking.setLength(currentText.length() - currentMarking.getStartPosition());
+                currentMarking = null;
             }
+
             // Add the token from this line to the document's text
+            if (whiteSpaceInFront && whiteSpaceBehindPreviousToken) {
+                currentText.append(whitespace);
+            } else {
+                currentText.append(nonWhitespace);
+            }
+            // If this line contains the start of a marking, we should keep track of it
+            if (token.length > annotationColumn && token[annotationColumn].startsWith(MARKING_START)) {
+                // Create new marking
+                currentMarking = createNewMarking(token, currentText.length());
+                markings.add(currentMarking);
+            }
             // TODO 1. make the whitespace configurable to allow other word separators 2.
             // Remove the previous word separator if we have a punctuation character.
             // (quotation, apostrophe)
-            currentText.append(token[0] + " ");
-            // Increase the line ID
-            i++;
+            currentText.append(normalizedToken);
+        }
+        // If there is an unfinished marking, finalize it
+        if (currentMarking != null) {
+            currentMarking.setLength(currentText.length() - currentMarking.getStartPosition());
         }
         return markings;
     }
 
-    protected Marking getWholeMarking(List<String> linesOfCurrentDoc, int pos, StringBuilder currentText) {
-        String[] tokens = linesOfCurrentDoc.get(pos).split("\t");
-
+    private Span createNewMarking(String[] line, int startPos) {
```

Comment:
1. A javadoc comment would be good.
2. Since this class is extended by other classes, it makes sense to have the visibility `protected` instead of `private`